File: nodes/logic/index_switch.py
"""Index Switch — pick the Nth connected input, with truly dynamic inputs.

The node starts with one input. The JS (web/index_switch.js) calls addInput
when the last slot gets connected and removes the trailing empties when you
disconnect — so the slot count grows/shrinks with what you actually wire.
No pre-declared ghost slots.

For that to work the backend must accept inputs it never declared: that's
what VALIDATE_INPUTS(**kwargs)->True does (the same escape hatch rgthree /
custom-scripts use). execute() catches every wired slot via **kwargs.

`index` selects among the connected inputs in slot order, and is CLAMPED to
range — out-of-range never errors (negative -> first, too big -> last).
"""

from __future__ import annotations

from ...base import TiNode
from ...registry import register


def _gather(kwargs: dict, prefix: str) -> list:
	"""Connected inputs (non-None) in numeric slot order: prefix1, prefix2…"""
	pairs = []
	for k, v in kwargs.items():
		if v is None or not k.startswith(prefix):
			continue
		try:
			n = int(k[len(prefix):])
		except ValueError:
			continue
		pairs.append((n, v))
	pairs.sort(key=lambda p: p[0])
	return [v for _, v in pairs]


def _clamp_pick(items: list, index: int):
	if not items:
		return None
	return items[max(0, min(int(index), len(items) - 1))]


@register
class ImageIndexSwitch(TiNode):
	DISPLAY_NAME = "Image Index Switch (ti)"
	CATEGORY = "tinode/logic"

	@classmethod
	def INPUT_TYPES(cls):
		# Only the first slot is declared; the JS adds image2, image3, … live.
		return {
			"required": {"index": ("INT", {"default": 0, "min": 0, "max": 9999})},
			"optional": {"image1": ("IMAGE",)},
		}

	@classmethod
	def VALIDATE_INPUTS(cls, **kwargs):
		# Accept the dynamically-added slots the backend never declared.
		return True

	RETURN_TYPES = ("IMAGE",)
	RETURN_NAMES = ("image",)
	FUNCTION = "execute"

	def execute(self, index=0, **kwargs):
		out = _clamp_pick(_gather(kwargs, "image"), index)
		if out is None:
			raise ValueError("Image Index Switch: no images connected")
		return (out,)


@register
class StringIndexSwitch(TiNode):
	DISPLAY_NAME = "String Index Switch (ti)"
	CATEGORY = "tinode/logic"

	@classmethod
	def INPUT_TYPES(cls):
		return {
			"required": {"index": ("INT", {"default": 0, "min": 0, "max": 9999})},
			"optional": {"string1": ("STRING", {"forceInput": True})},
		}

	@classmethod
	def VALIDATE_INPUTS(cls, **kwargs):
		return True

	RETURN_TYPES = ("STRING",)
	RETURN_NAMES = ("string",)
	FUNCTION = "execute"

	def execute(self, index=0, **kwargs):
		out = _clamp_pick(_gather(kwargs, "string"), index)
		return (out if out is not None else "",)

File: web/index_switch.js
// Truly dynamic inputs for the Index Switch nodes.
//
// The node ships with one slot (image1 / string1). When you connect the last
// slot, we addInput a fresh one; when you disconnect and leave trailing empty
// slots, we collapse them back to a single empty. So the slot count tracks
// what you actually wire — no pre-declared ghosts.
//
// Backend accepts the added slots via VALIDATE_INPUTS(**kwargs)->True.

import { app } from "../../scripts/app.js";

const NODES = {
	TI_ImageIndexSwitch: { prefix: "image", type: "IMAGE" },
	TI_StringIndexSwitch: { prefix: "string", type: "STRING" },
};

function dynList(node, prefix) {
	return (node.inputs || []).filter((i) => i && i.name && i.name.startsWith(prefix));
}

function update(node) {
	const { prefix, type } = node.__ti;

	// 1) collapse trailing empties down to a single one
	let guard = 0;
	while (guard++ < 500) {
		const dyn = dynList(node, prefix);
		if (dyn.length < 2) break;
		const last = dyn[dyn.length - 1];
		const prev = dyn[dyn.length - 2];
		if (last.link == null && prev.link == null) {
			node.removeInput(node.inputs.indexOf(last));
			continue;
		}
		break;
	}

	// 2) ensure exactly one empty trailing slot
	const dyn = dynList(node, prefix);
	const last = dyn[dyn.length - 1];
	if (!last || last.link != null) {
		let maxN = 0;
		for (const i of dyn) {
			const n = parseInt(i.name.slice(prefix.length), 10);
			if (n > maxN) maxN = n;
		}
		node.addInput(prefix + (maxN + 1), type);
	}

	node.setDirtyCanvas?.(true, true);
}

function safeUpdate(node, cfg) {
	node.__ti = cfg;
	try { update(node); } catch (e) { console.error("[tinode] indexSwitch", e); }
}

app.registerExtension({
	name: "tinode.indexSwitch",
	async setup() {
		console.log("[tinode] indexSwitch extension loaded");
	},
	beforeRegisterNodeDef(nodeType, nodeData) {
		const cfg = NODES[nodeData.name];
		if (!cfg) return;

		const onCreated = nodeType.prototype.onNodeCreated;
		nodeType.prototype.onNodeCreated = function () {
			onCreated?.apply(this, arguments);
			const self = this;
			requestAnimationFrame(() => safeUpdate(self, cfg));
		};

		const onConn = nodeType.prototype.onConnectionsChange;
		nodeType.prototype.onConnectionsChange = function () {
			onConn?.apply(this, arguments);
			safeUpdate(this, cfg);
		};
	},
});
